kafka-instrumentation/src/main/java/dev/responsive/kafka/clients/EndOffsetsPoller.java

```diff
@@ -0,0 +1,220 @@
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.serialization.BytesDeserializer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class polls the current end offset of the topic partitions assigned locally and
+ * publishes the values as a metric. It gets the set of assigned partitions by listening
+ * on ResponsiveConsumer.
+ *
+ */
+public class EndOffsetsPoller {
+  private static final Logger LOGGER = LoggerFactory.getLogger(EndOffsetsPoller.class);
+
+  private final Map<String, ThreadMetrics> threadIdToMetrics = new HashMap<>();
+  private final Metrics metrics;
+  private final Map<String, Object> configs;
+  private final Factories factories;
+  private ScheduledExecutorService executor;
+  private ScheduledFuture<?> pollFuture;
+  private Consumer<?, ?> consumer;
+
+  public EndOffsetsPoller(
+      final Map<String, ?> configs,
+      final Metrics metrics
+  ) {
+    this(configs, metrics, new Factories() {});
+  }
+
+  EndOffsetsPoller(
+      final Map<String, ?> configs,
+      final Metrics metrics,
+      final Factories factories
+  ) {
+    // TODO: make sure to use a read-committed consumer
+    this.configs = consumerConfigs(Objects.requireNonNull(configs));
+    this.metrics = Objects.requireNonNull(metrics);
+    this.factories = Objects.requireNonNull(factories);
+  }
+
+  private Map<String, Object> consumerConfigs(final Map<String, ?> providedConfigs) {
+    final HashMap<String, Object> configs = new HashMap<>(providedConfigs);
+    configs.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, BytesDeserializer.class);
+    configs.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, BytesDeserializer.class);
+    return Map.copyOf(configs);
+  }
+
+  public ResponsiveConsumer.Listener addForThread(final String threadId) {
+    LOGGER.info("add end offset metrics for thread {}", threadId);
+    final var tm = new ThreadMetrics(threadId);
+    synchronized (this) {
+      assert !threadIdToMetrics.containsKey(threadId);
+      if (threadIdToMetrics.isEmpty()) {
+        initPoller();
+        threadIdToMetrics.put(threadId, tm);
+      }
+    }
+    return tm;
+  }
+
+  public void removeForThread(final String threadId) {
+    final ThreadMetrics tm;
+    synchronized (this) {
+      tm = threadIdToMetrics.remove(threadId);
+      if (tm != null) {
+        tm.cleanup();
+        if (threadIdToMetrics.isEmpty()) {
+          stopPoller();
+        }
+      }
+    }
+  }
+
+  private void pollEndOffsets() {
+    final Set<TopicPartition> partitions = new HashSet<>();
+    final Collection<ThreadMetrics> threadMetrics;
+    synchronized (this) {
+      threadMetrics = List.copyOf(this.threadIdToMetrics.values());
+    }
+    for (final var tm : threadMetrics) {
+      partitions.addAll(tm.endOffsets.keySet());
+    }
+    final var endOffsets = consumer.endOffsets(partitions);
+    threadMetrics.forEach(tm -> tm.update(endOffsets));
+  }
+
+  private void initPoller() {
```

Comment:
can we also make this and `stopPoller` both synchronized methods? all access to them seems to be while holding `this` so it should be free and gives us an extra guarantee that they won't be called at the same time

Reply:
LMK if you still want this after the patch that cleans this up.